design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,293 @@
+
+For the example above, the `mysql-db-creds` variable will be saved in the database as follows:
+
+```json
+{
+  "username": "my-user",
+  "password": "U2FsdGVkX19Ji6JpgnVHW3V47OtMJwuKi1Yf9nc0aP5QcuzdnIrpzZ2zMC90f24g",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+Before the variable value is returned to the client, the secret attributes are decrypted so that the given example above, would look like this:
+
+```json
+{
+  "username": "my-user",
+  "password": "my-password",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+#### Built-in Schemas
+
+To simplify the user experience, Conjur can come with prdefined JSON schemas for the most common secrets use cases. The list of predefined schmeas should include:
+
+- Database secrets - Oracle, MSSQL, MySQL, Postgres, MariaDB, DB2.
+- Cloud access keys - AWS, Azure, GCP.
+- X509 certificates.
+- JWTs
+
+Write about the advantages of atomicity, minimal changes in the database to implement this and so on.
+
+#### Automatic Permission Granting to Built-in Schemas
+
+In order to provide access for any role to the built-in schemas, a new built-in group should be introduced: `conjur/all`. All roles in the Conjur account should be added to this group automatically, thus keeping this group always up to date with all the roles that exist in the account.
+
+Usage example:
+
+```yaml
+- !permit
+  role: !group /conjur/all
+  privileges: [ read,execute ]
+  resource: !variable my-var
+```
+
+## Main Advantages and Expected Value From This Proposal
+
+- We leverage a standard way to enforce the content of the variable, with a well known JSON schema.
+- Tightly coupled values, such as username and password, are updated together in a single transaction. This will prevent momentary inconsistency in which each value was updated independently, one after the other.
+- The user will not have to look in the documentation or to understand the required variables, their structure or allowed input. Conjur will provide this feedback when an update attempt is made.
+- The feature does not require changes in the database schma, only in new built-in content.
+- The returned variable is in a JSON structure, same as the rest of our APIs responses.
+
+## Affected Areas
+
+- The new functionality will be developed in the Conjur server and the policy parser.
+- No new APIs are needed, since all the functionality will be given as part of the policy loading.
+- No changes required in the database structure. But new built-in data should be added into the database tables, when they are initially created (clean install) or when an upgrade occurs from any prioir version.
```

Comment:
Se my UX comment, i think we will need to save the relation between the schemas to the variables in the DB

Reply:
Please see my response above.